{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fe18142c",
   "metadata": {},
   "source": [
    "# Parabricks: GPU-Accelerated Genomics, Ready for Local (WSL) and Cloud\n",
    "\n",
    "Genomic data volumes have exploded, but the *time* to turn raw reads into actionable variants no longer has to. **NVIDIA Parabricks** is a free software suite that accelerates secondary analysis of NGS DNA and RNA data—bringing end-to-end whole-genome workflows down from many hours on CPUs to minutes on GPUs, with outputs concordant to widely used tools. For example, Parabricks can analyze a 30× human whole genome in roughly **\\~10 minutes** on modern GPUs while matching the accuracy of trusted CPU pipelines. ([NVIDIA Docs][1])\n",
    "\n",
    "This repository, **`gabenavarro/parabricks`**, provides a practical, developer-friendly scaffold to run Parabricks:\n",
    "\n",
    "### Prerequisites\n",
    "\n",
    "---\n",
    "\n",
    "Before you begin, ensure you have the following installed on your local machine:\n",
    "\n",
    "- Docker: [Install Docker](https://docs.docker.com/get-docker/)\n",
    "- A compatible GPU\n",
    "- NVIDIA drivers\n",
    "\n",
    "\n",
    "## Building and Running the Docker Container\n",
    "\n",
    "---\n",
    "\n",
    "To build and run the Docker container for Parabricks, follow these steps:\n",
    "\n",
    "1. **Clone the Repository**: Clone the Parabricks repository to your local machine.\n",
    "\n",
    "```bash\n",
    "git clone https://github.com/gabenavarro/parabrick.git\n",
    "cd parabrick\n",
    "```\n",
    "\n",
    "2. **Build the Docker Image**: Use the provided Dockerfile to build the Docker image.\n",
    "\n",
    "```bash\n",
    "# You can choose any tag you want for the image\n",
    "# Feel free to play around with the base image, just make sure the host has the same or higher CUDA version\n",
    "docker build -f ./assets/build/Dockerfile -t parabricks:4.5.1-1 .\n",
    "```\n",
    "3. **Run the Docker Container**: Run the Docker container with the necessary configurations. In the first example, we will run the container locally with GPU support. This is the recommended way to run a container while in development mode. For scaling up, we will use the second example which runs the container in the cloud.\n",
    "\n",
    "```bash\n",
    "# Run the container with GPU support\n",
    "docker run -dt \\\n",
    "   --gpus all \\\n",
    "   --shm-size=128g \\\n",
    "   -v \"$(pwd):/workspace\" \\\n",
    "   --name parabricks \\\n",
    "   --env NVIDIA_VISIBLE_DEVICES=all \\\n",
    "   --entrypoint /bin/bash \\\n",
    "   parabricks:4.5.1-1\n",
    "```\n",
    "> Note: The `-v \"$(pwd):/workspace\"` option mounts the current directory to `/workspace` in the container, allowing you to access your local files from within the container. The `--env` options set environment variables for GPU visibility and Google Cloud credentials.<br>\n",
    "> Note: The `--gpus all` option allows the container to use all available GPUs. <br>\n",
    "\n",
    "4. **Access the Container with IDE**: In this example, we will use Visual Studio Code to access the container. You can use any IDE of your choice.\n",
    "\n",
    "```bash\n",
    "# In a scriptable manner\n",
    "CONTAINER_NAME=parabricks\n",
    "FOLDER=/workspace\n",
    "HEX_CONFIG=$(printf {\\\"containerName\\\":\\\"/$CONTAINER_NAME\\\"} | od -A n -t x1 | tr -d '[\\n\\t ]')\n",
    "code --folder-uri \"vscode-remote://attached-container+$HEX_CONFIG$FOLDER\"\n",
    "```\n",
    "\n",
    "> Note: The `code` command is used to open Visual Studio Code. Make sure you have the Remote - Containers extension installed in VS Code to access the container directly. <br>\n",
    "> Note: Make sure you have installed Remote - Containers extension in VS Code.<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c9d2c93",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
